### PR TITLE
Add alternatives section to beforematch explainer

### PR DESCRIPTION
**Pull request by josepharhar**
I made example one more complicated in order to make some points i the alternatives, I hope thats OK

### CODE REVIEW

File: explainers/beforematch.md [hunk 1]
```diff
@@ -288,6 +307,50 @@ scroll after adding an async step.
...
+
+#### Automatic Unlocking with Activation Event
+This is like "Automatic Unlocking," but with an added "activation" emitted when
+content is displayed to allow developers to change other state and styles if
+needed.
+##### Pros
+* Same pros as "Automatic Unlocking."
+* Allows other state/styles to be changed when content is displayed.
+##### Cons
+* TODO: ask vmpstr and chrishtr for more context on this
```
Comment (josepharhar) [inline explainers/beforematch.md:335]: What was the reason not to go with activation events again?

Comment (vmpstr) [reply explainers/beforematch.md:335]: So it depends how you look at it
- On the one hand, activation event goes hand-in-hand with automatic unlocking/expanding, which we think is not a good use-case since the majority of the time the developer would want to do something other than just show the element. That, and it makes state tracking hard (ie the developer doesn't have the tools to collapse the section again; we would have to add something like a JS api). In this case activation event informs that this has happened. However, since we're not doing the automatic unlocking then activation event is not needed.

- On the other hand, activation event could be fired prior to us finding a match that would have otherwise unlocked the element. That's exactly what beforematch is :) In other words, we renamed activation event to beforematch when we decided not to automatically unlock due to reasoning above.

Does that make sense?

File: explainers/beforematch.md [hunk 2]
```diff
@@ -288,6 +307,50 @@ scroll after adding an async step.
 5. **content-visibility: hidden-matchable -> visible**: find-in-page scrolls to
 to the revealed text.
 
+### Alternatives
+Given the purpose of displaying `content-visibility: hidden-matchable` text
+when it is searched for, there are a number of alternatives we have considered.
+
+#### Automatic Unlocking
+`content-visibility: hidden-matchable` text would be automatically changed to
+`content-visibility: visible` when the hidden text is searched for.
```
Comment (vmpstr) [inline explainers/beforematch.md:316]: This is not actually correct. We would keep an internal state that essentially ignores 'hidden-matchable'. The problem is that, unlike attributes, we can't change styles from within the browser. 

If we have something like
```
body > .randomClass {
  content-visibility: hidden-matchable;
}

it means that all first level children of body that have .randomClass have hidden-matchable. If we find one such element, we obviously can't change this style rule since it doesn't just apply to this element. We could change inline style of the element, but that sounds like a really bad approach since we don't know how script is using inline styles, so we shouldn't be messing with them.

File: explainers/beforematch.md [hunk 3]
```diff
@@ -288,6 +307,50 @@ scroll after adding an async step.
...
+##### Pros
+* Provides the desired behavior without the need for the web developer to do
+  anything besides use the `hidden-matchable` value.
+* Less complex than `beforematch` with regards to scrolling behavior.
```
Comment (vmpstr) [inline explainers/beforematch.md:320]: We need to elaborate on this: Specifically, since script does not run and the interaction is fully within browser control, we can guarantee that scrolling to the element is possible

File: explainers/beforematch.md [hunk 4]
```diff
@@ -288,6 +307,50 @@ scroll after adding an async step.
...
+##### Cons
+* Doesn't allow the developer to change other properties in conjunction with
```
Comment (vmpstr) [inline explainers/beforematch.md:322]: One other con is what I said before: we need to have some internal state tracking that isn't obvious

File: explainers/beforematch.md (hunk 1 above)
Comment (vmpstr) [reply explainers/beforematch.md:335]: Just to add to this as I'm reading the doc, the important con is that automatically changing CSS isn't possible, so there is a lot of complexity involved with automatic unlocking. And the whole point of this activation event here is that if we're firing an activation event anyway, then why bother with automatic unlocking? Just let the script take care of it and reduce the complexity and internal states in the browser.

To put it differently, automatic unlocking with an event is not any better than just the event, because the handler in the event could still remove the element, relock the element, move it, etc, etc. All of those need to be dealt with whether or not we automatically unlock in the presence of an event.

File: explainers/beforematch.md [hunk 5]
```diff
@@ -288,6 +307,50 @@ scroll after adding an async step.
...
+* Same pros as "Automatic Unlocking."
+* Allows other state/styles to be changed when content is displayed.
+##### Cons
+* TODO: ask vmpstr and chrishtr for more context on this
+
+#### CSS Pseudo Selector
+A pseudo selector, such as `:target`, would be applied to the element
+containing the matching text when it is searched for.
+##### Pros
+* Allows content to become visible when searched for with only CSS.
```
Comment (vmpstr) [inline explainers/beforematch.md:341]: Also allows custom styling on the element

File: explainers/beforematch.md [hunk 6]
```diff
@@ -288,6 +307,50 @@ scroll after adding an async step.
...
+needed.
+##### Pros
+* Same pros as "Automatic Unlocking."
+* Allows other state/styles to be changed when content is displayed.
+##### Cons
+* TODO: ask vmpstr and chrishtr for more context on this
+
+#### CSS Pseudo Selector
+A pseudo selector, such as `:target`, would be applied to the element
+containing the matching text when it is searched for.
```
Comment (vmpstr) [inline explainers/beforematch.md:339]: Also possibly the same or different selector on the ancestor chain to allow for different styling

File: explainers/beforematch.md [hunk 7]
```diff
@@ -288,6 +307,50 @@ scroll after adding an async step.
...
+* Allows content to become visible when searched for with only CSS.
+##### Cons
+* If CSS with a pseudo selector is used to make text visible, then when
+  find-in-page is closed or the search text changes, the pseudo selector would
+  be removed and then any selector which is displaying the text based on that
+  pseudo selector would not apply, causing the expanded section to unexpectedly
+  collapse.
+* There is no way in CSS to say "change my style if a descendant has a pseudo
+  class on it." For this reason, developers would be unable to change styles
+  outside of the particular element that has the matching text, which would
```
Comment (vmpstr) [inline explainers/beforematch.md:350]: This isn't necessarily true if we add the pseudo class on the ancestor hierarchy. I think the point here is that it can be complicated or impossible to provide the right selector which can modify a style on an unrelated element (Such as the case in the title-collapsed or title-expanded that you have above

File: explainers/beforematch.md [hunk 8]
```diff
@@ -288,6 +307,50 @@ scroll after adding an async step.
...
+##### Cons
+* TODO: ask vmpstr and chrishtr for more context on this
+
+#### CSS Pseudo Selector
+A pseudo selector, such as `:target`, would be applied to the element
+containing the matching text when it is searched for.
+##### Pros
+* Allows content to become visible when searched for with only CSS.
+##### Cons
+* If CSS with a pseudo selector is used to make text visible, then when
```
Comment (vmpstr) [inline explainers/beforematch.md:343]: Is it possible for you to take a few screenshots on a phone interaction with mobile wiki page
- One screen of collapsed sections
- One screen of find-in-page that finds text in a collapsed section (you can manually expand the section for the purpose of the demo)
- One screen with find-in-page dialog dismissed that still displays the open section.

I think that example would make this con point obvious

Comment (vmpstr) [reply explainers/beforematch.md:343]: Nevermind, I see you added a similar thing in the #172

File: explainers/beforematch.md (hunk 1 above)
Comment (josepharhar) [reply explainers/beforematch.md:335]: Why isn't automatically changing CSS possible? If we are able to fire an event that has the power to change CSS, then can't we just change the style from `hidden-matchable` to `visible` in C++?

File: explainers/beforematch.md (hunk 3 above)
Comment (josepharhar) [reply explainers/beforematch.md]: I reworded this bullet point. Look good?

File: explainers/beforematch.md (hunk 5 above)
Comment (josepharhar) [reply explainers/beforematch.md:365]: Done

File: explainers/beforematch.md (hunk 6 above)
Comment (josepharhar) [reply explainers/beforematch.md:363]: done

File: explainers/beforematch.md (hunk 1 above)
Comment (vmpstr) [reply explainers/beforematch.md:359]: Which style do we change? We can't change the rule style because the rule applies to more than one element (in the general case; ie * { content-visibility: hidden-matchable } applies to _all_ elements, so when we activate, we can't just change that to visible since that un-hides everything). 

The only possibility is that we set the inline style of the element to have content-visibility: visible. But again, if for example the descendant elements have content-visibility: inherit or something then we're changing more than just the one element. That, and changing the inline style automatically open the door to a lot of web breaking, since if script generates and sets a new inline style, all of a sudden it overrides the values we set. That's just not going to be a good change.

The idea of the event is to defer this decision to script, since there is no one correct decision in terms of what to do, so we have to delegate the decision to the app.

File: explainers/beforematch.md (hunk 3 above)
Comment (vmpstr) [reply explainers/beforematch.md]: Yep, this looks good.

File: explainers/beforematch.md (hunk 1 above)
Comment (josepharhar) [reply explainers/beforematch.md:359]: Ah I see, having the browser change styles doesn't really make sense. Thanks for the explanation.
I added another alternative to show why this is a bad idea.

File: explainers/beforematch.md (hunk 2 above)
Comment (josepharhar) [reply explainers/beforematch.md:338]: Thanks for the explanation, I resolved this in another thread

File: explainers/beforematch.md (hunk 4 above)
Comment (josepharhar) [reply explainers/beforematch.md:346]: Done

File: explainers/beforematch.md (hunk 7 above)
Comment (josepharhar) [reply explainers/beforematch.md:376]: Thanks for the explanation, I put your words into this bullet point
